app/src/main/kotlin/com/wire/android/ui/home/settings/backup/BackupAndRestoreViewModel.kt

```diff
@@ -2,118 +2,188 @@ package com.wire.android.ui.home.settings.backup
+            } else {
+                // If it's not password protected, we can restore the backup
+                val result = importBackup(extractedBackupFilesRootPath, null)
+                if (result is RestoreBackupResult.Success) {
+                    Toast.makeText(context, context.getString(R.string.backup_label_conversation_successfully_saved), Toast.LENGTH_LONG)
+                        .show()
+                } else {
+                    onBackupRestoreError("Error when restoring the db file", RestoreFileValidation.IncompatibleBackup)
+                }
+            }
+        })
     }
 
-    private fun validateBackupFile(uri: Uri) {
-        state = state.copy(restoreFileValidation = RestoreFileValidation.PasswordRequired)
+    private fun onBackupRestoreError(errorMessage: String, errorType: RestoreFileValidation = RestoreFileValidation.GeneralFailure) {
+        appLogger.e(errorMessage)
+        state = state.copy(restoreFileValidation = errorType)
     }
 
-    fun restoreBackup(restorePassword: TextFieldValue) {
-        viewModelScope.launch {
-            //TODO: restore the back up file
-            checkRestorePassword(restorePassword)
-            //TODO: restore the file
-            restoreBackupFile(restorePassword)
-        }
-    }
     @Suppress("MagicNumber")
-    private suspend fun restoreBackupFile(restorePassword: TextFieldValue) {
-        // Test purpose remove once we are able to restore file
-        state = state.copy(restoreProgress = RestoreProgress.InProgress(0.25f))
-        delay(250)
-        state = state.copy(restoreProgress = RestoreProgress.InProgress(0.50f))
-        delay(250)
-        state = state.copy(restoreProgress = RestoreProgress.InProgress(0.75f))
-        delay(250)
-        state = state.copy(restoreProgress = RestoreProgress.InProgress(0.99f))
+    fun restorePasswordProtectedBackup(restorePassword: TextFieldValue) = viewModelScope.launch {
+        state = state.copy(
+            backupRestoreProgress = BackupRestoreProgress.InProgress(0.50f),
+            restorePasswordValidation = PasswordValidation.NotVerified
+        )
         delay(250)
-        state = state.copy(restoreProgress = RestoreProgress.Finished)
+        val fileValidationState = state.restoreFileValidation
+        if (fileValidationState is RestoreFileValidation.PasswordRequired) {
+            when (val result = importBackup(fileValidationState.extractedBackupFilesRootPath, restorePassword.text)) {
+                RestoreBackupResult.Success -> {
+                    state = state.copy(
+                        backupRestoreProgress = BackupRestoreProgress.Finished,
+                        restorePasswordValidation = PasswordValidation.Valid
+                    )
+                    Toast.makeText(context, context.getString(R.string.backup_label_conversation_successfully_saved), Toast.LENGTH_LONG)
```

Comment:
we should show toast and stuff from the UI, this is a rule we kind of agreed on, also if we do it this way, the toast will not react for the language change for example if the user decides to change I know a really an edge case. 

Also I am not 100 % sure but it could cause a memory leak when you show the Toast but the user navigates away from the Composalbe holding ref to this ViewModel, if Toast decides to live longer than Composable, which it probably will, the ViewModel will not be garbage collected because Toast is holding a ref to context, which is being hold by a ViewModel, so moving this to UI would minimise any drawbacks

Reply:
this was just a leftover from a test session I did when implementing initially. I would move this Toast to UI screen or even better, implement it with Snackbar. Thanks for pointing it out tho :)